internal/tofu/upgrade_resource_state.go

```diff
@@ -18,82 +18,143 @@ import (
+		log.Printf("[TRACE] upgradeResourceStateTransform: failed - address: %s", args.currentAddr)
+		return cty.NilVal, nil, diags
 	}
 
-	resp := provider.UpgradeResourceState(req)
+	return resp.UpgradedState, args.objectSrc.Private, diags
+}
+
+// moveResourceState moves the state from one type or provider to another
+// this function runs the provider-defined logic for MoveResourceState and performs the additional validation defined int transformResourceState
+// If any error occurs during the validation or state transformation, error diagnostics are returned.
+// Otherwise, the new state object is returned.
+func moveResourceState(params stateTransformArgs) (*states.ResourceInstanceObjectSrc, tfdiags.Diagnostics) {
+	return transformResourceState(params, moveResourceStateTransform)
+}
+
+// moveResourceStateTransform is a providerStateTransform that moves the state via provider move logic
+func moveResourceStateTransform(args stateTransformArgs) (cty.Value, []byte, tfdiags.Diagnostics) {
+	log.Printf("[TRACE] moveResourceStateTransform: new address: %s, previous address: %s", args.currentAddr, args.prevAddr)
+	req := providers.MoveResourceStateRequest{
+		SourceProviderAddress: args.prevAddr.Resource.Resource.ImpliedProvider(),
+		SourceTypeName:        args.prevAddr.Resource.Resource.Type,
+		SourceSchemaVersion:   args.objectSrc.SchemaVersion,
+		SourceStateJSON:       args.objectSrc.AttrsJSON,
+		SourceStateFlatmap:    args.objectSrc.AttrsFlat,
+		SourcePrivate:         args.objectSrc.Private,
+		TargetTypeName:        args.currentAddr.Resource.Resource.Type,
+	}
+	resp := args.provider.MoveResourceState(req)
 	diags := resp.Diagnostics
+	if diags.HasErrors() {
+		log.Printf("[TRACE] moveResourceStateTransform: failed - new address: %s, previous address: %s - diags: %s", args.currentAddr, args.prevAddr, diags.Err().Error())
+		return cty.NilVal, nil, diags
+	}
+	return resp.TargetState, resp.TargetPrivate, diags
+}
+
+// transformResourceState transforms the state based on passed in providerStateTransform and returns new ResourceInstanceObjectSrc
+// the function takes in the current and previous AbsResourceInstance, Provider, ResourceInstanceObjectSrc, current schema and current version
+func transformResourceState(args stateTransformArgs, stateTransform providerStateTransform) (*states.ResourceInstanceObjectSrc, tfdiags.Diagnostics) {
+	if args.currentAddr.Resource.Resource.Mode != addrs.ManagedResourceMode {
+		// We only do state transformation for managed resources.
+		// This was a part of the normal workflow in older versions and
+		// returned early, so we are only going to log the error for now.
+		log.Printf("[ERROR] data resource %s should not require state transformation", args.currentAddr)
+		return args.objectSrc, nil
+	}
+
+	// TODO: This should eventually use a proper FQN.
+	providerType := args.currentAddr.Resource.Resource.ImpliedProvider()
+	if args.objectSrc.SchemaVersion > args.currentSchemaVersion {
```

Comment:
I don't think this check makes sense for moved.  The schema version between resources/providers during a move has no bearing here.  This should be moved to the upgrade case only